### PR TITLE
Enforce worker limits in Compression Streams API

### PR DESCRIPTION
**Pull request by fhanau**

### CODE REVIEW

**Comment by fhanau** [comment]
This is modeled after the limiting approach recently added in dh.c++ and should avoid CPU utilization issues when compressing a large amount of data at once.
Some things I'm not sure about:
1) This assumes that querying the limiter is somewhat expensive. Hence, we try to avoid it if the output data fits into the output buffer, so that we don't need to query this every time when compressing small amounts of data/compressing data in small chunks which should already get limited properly. Is there actually such overhead, for example from having to acquire locks?
2) Is this the right spot for this to apply exception handling correctly?

**Review by anonrig — COMMENTED** [review]
Is this easily testable? I feel such that should have a test that validates the behavior.

**Comment by jasnell** [comment]
Testing this in workerd is likely difficult since we do not actually implement the cpu limiting functionality (it's all a non-op in workerd). We'll likely need to add a test to the internal repo separately.

**Comment by jasnell** [comment]
Labeling as "needs-internal-pr" so that we can get a test added internally to verify this

File: src/workerd/api/streams/compression.c++ [hunk 1]
```diff
@@ -342,6 +349,18 @@ private:
...
       KJ_IF_SOME(exception, kj::runCatchingExceptions([this, flush, &result]() {
         result = context.pumpOnce(flush);
+        // we only want to apply limit checks on calls where there may be a large amount of data
+        // being processed. Only check limits when we have filled the entire output buffer,
+        // indicating that there is more data to be (de)compressed.
+        if (result.buffer.size() == ZLIB_OUT_BUF_SIZE) {
+          KJ_IF_SOME(outcome, IoContext::current().getLimitEnforcer().getLimitsExceeded()) {
+            if (outcome == EventOutcome::EXCEEDED_CPU) {
+              JSG_FAIL_REQUIRE(Error, "Compression Stream write failed: Exceeded CPU limit");
+            } else if (outcome == EventOutcome::EXCEEDED_MEMORY) {
```
Comment (jasnell) [inline src/workerd/api/streams/compression.c++:359]: So one key bit here is that I don't believe we are yet reporting the size of the buffer here to v8, so v8 itself won't really have any insight into the memory consumed by the decompression itself -- that is, unless these chunks are immediately being moved into a `v8::ArrayBuffer`. So I'm not sure if the memory check here will be effective? Have you been able to verify this appropriately accounts for the memory used during decompression?

**Comment by fhanau** [comment]
Downstream PR and test have been added.

File: src/workerd/api/streams/compression.c++ (hunk 1 above)
Comment (fhanau) [reply src/workerd/api/streams/compression.c++:359]: This change was designed with CPU time limiting in mind; it does not take any additional steps to accurately measure memory consumption, but once that is implemented properly it should be able to limit that too. I can add a comment indicating this.

**Review by jasnell — APPROVED** [review]

**Review by anonrig — APPROVED** [review]

**Pull request merged**